File: library/src/main/java/com/example/library/mapper/BookMapper.java
package com.example.library.mapper;

import java.time.LocalDateTime;

import com.example.library.dto.request.BookRequest;
import com.example.library.dto.request.CategoryRequest;
import com.example.library.dto.response.BookResponse;
import com.example.library.dto.response.BookSyncResponse;
import com.example.library.entity.Account;
import com.example.library.entity.Book;
import com.example.library.entity.Category;

public class BookMapper {

    public static BookSyncResponse toSyncResponse(Book book) {
        BookSyncResponse response = new BookSyncResponse();

        response.setId(book.getId());
        response.setBookCode(book.getBookCode());
        response.setBookName(book.getBookName());
        
        response.setPrice(book.getPrice());
        response.setCreatedBy(book.getCreatedBy().getId());
        response.setCategoryId(book.getCategory().getId());

        response.setUpdatedAt(book.getUpdatedAt());
        response.setDeleteAt(book.getDeleteAt());
        response.setVersion(book.getVersion());

        return response;
    }
    

    public static BookResponse toBookResponse(Book book) {

        return BookResponse.builder()
                .id(book.getId())
                .bookCode(book.getBookCode())
                .bookName(book.getBookName())
                .price(book.getPrice())
                .categoryName(book.getCategory().getCategoryName())
                .createdBy(book.getCreatedBy().getUsername())
                .categoryId(book.getCategory().getId())
                .updatedAt(book.getUpdatedAt())
                .deleteAt(book.getDeleteAt())
                .build();
    }
   
    public static Book toBook(BookRequest request, Account account,Category category ) {

        return Book.builder()
                .bookCode(request.getBookCode())
                .bookName(request.getBookName())
                .price(request.getPrice())
                .version(0L)  
                .updatedAt(LocalDateTime.now())
                .deleteAt(null)
                .category(category)
                .createdBy(account)
                .build();
    }
}

File: MyApplication/app/src/main/java/com/example/myapplication/data/mapper/BookMapper.kt
package com.example.myapplication.data.mapper

import com.example.myapplication.data.local.entity.BookEntity
import com.example.myapplication.data.remote.dto.response.BookResponse
import com.example.myapplication.data.remote.dto.response.BookSyncResponse
import com.example.myapplication.data.remote.dto.response.LoginResponse
import com.example.myapplication.domain.model.Book
import com.example.myapplication.domain.model.LoginResult
import com.example.myapplication.feature.librarian.book.BookEvent
import kotlin.Int

object bookMapper {
    fun BookResponse.toBook(): Book {
        return Book(id,bookCode,bookName,price, categoryId)
    }
//    fun BookResponse.toEntiry(): BookEntity {
//        return BookEntity(id,bookCode,bookName,price, categoryId,updatedAt,deleteAt,accountId, version = )
//    }
    fun BookEntity.toBook(): Book {
        return Book(id,bookCode,bookName,price, categoryId)
    }
    fun BookSyncResponse.toEntity(): BookEntity {
        return BookEntity(
            id = id,
            bookCode = bookCode,
            bookName = bookName,
            price = price,
            categoryId = categoryId,
            updatedAt = updatedAt,
            deleteAt = deleteAt,
            accountId = createdBy,
            version = version
        )
    }
}